clients/src/main/java/org/apache/kafka/clients/consumer/internals/ConsumerCoordinator.java

```diff
@@ -362,25 +363,63 @@ public void commitOffsetsSync(Map<TopicPartition, OffsetAndMetadata> offsets) {
-            public void run(long now) {
-                commitOffsetsAsync(subscriptions.allConsumed(), new OffsetCommitCallback() {
-                    @Override
-                    public void onComplete(Map<TopicPartition, OffsetAndMetadata> offsets, Exception exception) {
-                        if (exception != null)
-                            log.error("Auto offset commit failed.", exception);
-                    }
-                });
-                client.schedule(this, now + interval);
+    private class AutoCommitTask implements DelayedTask {
+        private final long interval;
+        private boolean enabled = false;
+
+        public AutoCommitTask(long interval) {
+            this.interval = interval;
+        }
+
+        public void enable() {
+            if (!enabled) {
+                // there shouldn't be any instances scheduled, but call unschedule anyway to ensure
+                // that this task is only ever scheduled once
+                client.unschedule(this);
+                this.enabled = true;
+                long now = time.milliseconds();
+                client.schedule(this, interval + now);
+            }
+        }
+
+        public void disable() {
+            this.enabled = false;
+            client.unschedule(this);
+        }
+
+        public void run(final long now) {
+            if (!enabled)
+                return;
+
+            if (coordinatorUnknown()) {
+                log.debug("Cannot auto-commit offsets now since the coordinator is unknown, will retry after backoff");
+                client.schedule(this, now + retryBackoffMs);
+                return;
             }
-        };
-        client.schedule(task, time.milliseconds() + interval);
-        return task;
+
+            commitOffsetsAsync(subscriptions.allConsumed(), new OffsetCommitCallback() {
+                @Override
+                public void onComplete(Map<TopicPartition, OffsetAndMetadata> offsets, Exception exception) {
+                    if (exception == null) {
+                        client.schedule(AutoCommitTask.this, now + interval);
+                    } else if (exception instanceof RetriableException) {
+                        log.debug("Auto offset commit failed, will retry after backoff.", exception);
+                        client.schedule(AutoCommitTask.this, time.milliseconds() + retryBackoffMs);
+                    } else {
+                        log.warn("Auto offset commit failed: {}", exception.getMessage());
+                        client.schedule(AutoCommitTask.this, now + interval);
+                    }
+                }
+            });
+        }
     }
 
     private void maybeAutoCommitOffsetsSync() {
         if (autoCommitEnabled) {
+            // disable periodic commits prior to committing synchronously. note that they will
+            // be re-enabled after a rebalance completes
+            autoCommitTask.disable();
```

Comment:
On second thought, there might be an issue here. The problem is what happens if the task is disabled and re-enabled while an offset commit is in-flight. When the in-flight request returns, it would find the task still enabled and would reschedule. We might also have this problem for heartbeats. One way to solve it might be to add an inFlightRequest flag and skip scheduling in enable() when it is set.